### PR TITLE
profiler: Deprecate WithAPIKey even harder

### PR DESCRIPTION
**Pull request by felixge**
@AlexJF can you take a look at this suggested API change for PROF-2719 and confirm that it is what you had in mind? I haven't hooked it up in the code yet and wanted to make sure you're happy with the naming and customer facing documentation. I think this would do a better job at communicating intend than the `WithAgentlessUpload` option we discussed.

### CODE REVIEW

File: profiler/options.go [hunk 1]
```diff
@@ -197,15 +198,28 @@ func WithAgentAddr(hostport string) Option {
...
-// using the provided API key.
+// WithAPIKey is deprecated and will be ignored unless you also use the
+// WithDeprecatedAPIKey() option which has more information on the deprecation.
 func WithAPIKey(key string) Option {
 	return func(cfg *config) {
 		cfg.apiKey = key
 	}
 }
 
+// WithDeprecatedAPIKey exists to enable the deprecated agentless upload mode.
```
Comment (felixge) [inline profiler/options.go:209]: We could also call this `WithDeprecatedAgentlessMode` or something. I just think `WithAgentlessUpload` doesn't make this option sound scary enough to raise red flags with customers.

File: profiler/options.go [hunk 2]
```diff
@@ -197,15 +198,28 @@ func WithAgentAddr(hostport string) Option {
 	}
 }
 
-// WithAPIKey is deprecated and might be removed in future versions of this
-// package. It allows to skip the agent and talk to the Datadog API directly
-// using the provided API key.
+// WithAPIKey is deprecated and will be ignored unless you also use the
```
Comment (gbbr) [inline profiler/options.go:201]: This is a breaking change IMHO. Removing this option vs. doing this - it's basically the same thing. I would ask you to consider logging a warning instead.

File: profiler/options.go (hunk 1 above)
Comment (gbbr) [reply profiler/options.go:209]: Why would you add a new option which is deprecated straight out of the box? I'm pretty sure you guys talked about this forehand and I'm out of context, but just putting in an outsider perspective.

File: profiler/options.go (hunk 2 above)
Comment (felixge) [reply profiler/options.go:201]: @gbbr yes, it's a breaking change. There is more discussion about this in `PROF-2719`. 

The backend might eventually reject agentless-profile uploads, so this breaking change will happen regardless of making this change in the library or not.

IDK what semver has to say about "backend API changes", but clearly those kind of changes can often be outside of the control of the library.

I think this is the most user-friendly way to deal with this that won't force an immediate v2 release of dd-trace-go.

PS: [Everything](https://www.hyrumslaw.com/) is a breaking change 🙈 and semver doesn't have a lot to say about non-obvious breaking changes like this one. So for better or worse, every single change we make is a subtle tradeoff between "strict semver" less "toil" : /.

Comment (felixge) [reply profiler/options.go:201]: Oh, this is a fun watch if you want to look further into the abyss of Hyrum's Law and how another part of Google has a totally different approach to tackling it compared to the Go/Semver folks: [CppCon 2017: Titus Winters “C++ as a "Live at Head" Language” - YouTube](https://www.youtube.com/watch?v=tISy7EJQPzI)

File: profiler/options.go (hunk 1 above)
Comment (felixge) [inline profiler/options.go:209]: Note to self: Rename to `WithAgentlessUpload` as discussed with @AlexJF

File: profiler/options.go (hunk 2 above)
Comment (gbbr) [reply profiler/options.go:201]: What you're doing here:
* Making an existing (deprecated) option a no-op.
* Adding a new (deprecated out of the box) option to make the previous no-op option work.
What I'm proposing:
* Keep the original option
* Drop the new one
* Log a warning.

**Comment by felixge** [comment]
Closing this in favor of https://github.com/DataDog/dd-trace-go/pull/893 (don't review yet)

**Pull request closed without merging**